provisioning/baremetal_config.go

```diff
@@ -75,56 +78,129 @@ func getProvisioningNetworkMode(prov *metal3iov1alpha1.Provisioning) metal3iov1a
 }
 
 func validateManagedConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningInterface", Value: prov.Spec.ProvisioningInterface},
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningDHCPRange", Value: prov.Spec.ProvisioningDHCPRange},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
-		}
+	provisionConfig := map[string]string{
```

Comment:
What is the benefit of copying data out of well-defined types into maps like this?

Reply:
Thanks for the quick review Doug!  Well there are a few options but I think this one produces the least amount of code.  Basically this just informs us what variables we are interested in testing.  You could use another method to do this but if you want to do it all in one function, as I have done here, you need a way to know what variables are being checked.

The reason this helps is because provisioning IP, DHCP range and network CIDR are all checked against each other.  So we need to know which are important and need to be checked and which can be ignored.  Otherwise we'd have to split them all up and somehow demarcate the involved variables.

This just seemed to me to be the best way to approach this.  I'm open to other ideas.

Reply:
I should say the current version allows us to test all states with a single function just by changing what keys are set in the map.

Reply:
Couldn't we also achieve that by looking directly at the Provisioning resource in the validation functions?

Reply:
I guess if we did some validation in each of validateManagedConfig(), validateUnmanagedConfig() and validateDisabledConfig to ensure that each only sets the variables we need then we could do it.  If a user wanted to move to Disabled config for example but left the DHCPRange set.. I'm not sure how we'd deal with that.  Error out?  Set it to "" for them?

Reply:
Why do we care if they leave extra data defined? Why not just apply the rules based on the mode that is set?

Reply:
then we have to know the mode.. I guess that's another way to do it.  Pass in the mode to the validations.

Reply:
Would it make sense to have a separate validation function for each mode, dispatched from a top level function? How different are the validations in each mode?

Reply:
There is a large amount of overlap.  The largest has 5 fields to validate while the smallest has a subset of 3 of those fields.  And again, they are interdependent which is why I took this approach.  It would be a mess to do them all separately.

Reply:
So far I see this approach, or passing in the mode as the best options.

Reply:
Isn't the mode part of the data in the Provisioning resource, though? Or at least derivable from it? Why do we need to pass it as an argument?

Reply:
Expanding the collapsed part of the diff, I see that we do have a single top level function dispatching to mode-specific functions at this level, which then put some of the configuration values into a map and passes that to another single function that is named generically but actually performs some mode-specific checks based on which values are found in the map. So, maybe passing the mode into validateProvisioningConfig makes sense.

Alternatively, creating a bunch of little functions for validating the parts of the config and then invoking each of those from these mode-specific functions would make sense and give clear, testable, and reusable code, without obfuscating the mode-specific behavior behind a map.